### ISSUE TITLE
Proposal: ref struct ConditionalWeakTable<TKey, TValue>.GetEnumerator()

### ISSUE THREAD

**Issue by Sergio0694** [dotnet/runtime#53296]
## Background and Motivation

The [`ConditionalWeakTable<TKey, TValue>`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.conditionalweaktable-2) is an extremely useful type that allows developers to leverage GC dependent handles (which are not directly available), along with some other nice features on top (eg. auto trimming, thread-safety by default, etc.). There is currently one drawback though when trying to use this type in code that's trying to be as fast and memory efficient as possible, which is that it is not possible to enumerate key-value pairs in a `ConditionalWeakTable<TKey, TValue>` table without incurring in unwanted allocations and extra overhead. Specifically:
- The [`GetEnumerator`](https://github.com/dotnet/runtime/blob/5525d0bbb57d5550d71a0a15158f992b0e694000/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/ConditionalWeakTable.cs#L235) implementation returns a reference type enumerator, incurring in one allocation
- The [enumerator type](https://github.com/dotnet/runtime/blob/5525d0bbb57d5550d71a0a15158f992b0e694000/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/ConditionalWeakTable.cs#L289) also has a finalizer, which makes the allocation slower even though it's not actually needed in most cases (given that when using `foreach`, `Dispose()` is always called anyway, which suppresses the finalizer).

## Proposed API

This proposal is to expose a value-based enumerator, allowing developers to enumerate key-value pairs in `ConditionalWeakTable<TKey, TValue>` instances without incurring in any allocations at all, and with also some nice performance improvements on top (which are just a natural consequence of moving to a value-based enumerator in this case and removing the two performance penalties mentioned before).

The proposed API is this one:

```diff
namespace System.Runtime.CompilerServices
{
    public sealed class ConditionalWeakTable<TKey, TValue> : IEnumerable<KeyValuePair<TKey, TValue>>
        where TKey : class
        where TValue : class?
    {
+       public Enumerator GetEnumerator();
        
+       public ref struct Enumerator
+       {
+           public KeyValuePair<TKey, TValue> Current { get; }
+
+           public bool MoveNext();
+           public void Dispose();
+       }
    }
}
```

This is perfectly safe when used through the `foreach` pattern, given that the C# compiler will still generate the same calls as when using a normal iterator, with the same `try/finally` blocks and the final `Dispose()` call (which is needed here to decrement the internal reference count for the `ConditionalWeakTable<TKey, TValue>` instance).

I'd be happy to contribute this feature myself in case the proposal was approved 🙌

## Usage Examples

We're using `ConditionalWeakTable<TKey, TValue>` in the [`WeakReferenceMessenger`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.toolkit.mvvm.Messaging.WeakReferenceMessenger) type in the [MVVM Toolkit](https://docs.microsoft.com/en-us/windows/communitytoolkit/mvvm/introduction). Currently this type incurs in allocations when broadcasting messages solely for the `Enumerator` type being instantiated every single time. As a test, I've hacked the `DependentHandle` type in a benchmark project (by defining it in a dummy `System.Private.CoreLib` project and using `[IgnoreAccessChecksTo]` and `PrivateAssets="All"` so that the MVVM Toolkit would load the true one from CoreCLR instead, [see here](https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3367), and then I copied the `ConditionalWeakTable<TKey, TValue>` source to it while adding the new API mentioned above to it. I then proceeded to run the same benchmark ([see here](https://gist.github.com/Sergio0694/b9dbf42bcf2fd3a412c7d3ff8d2b6de9)) across the original implementation of `WeakReferenceMessenger`, and one using this modified `ConditionalWeakTable<TKey, TValue>` implementation:

|   Method |     Mean |    Error |   StdDev | Ratio | Allocated |
|--------- |---------:|---------:|---------:|------:|----------:|
| Original | 21.79 ms | 0.296 ms | 0.277 ms |  1.00 | 192,029 B |
| Improved | **18.66 ms** | 0.216 ms | 0.191 ms |  **0.86** |         **-** |

✅ Completely allocation free broadcast
✅ About 15% faster too while at it, which doesn't hurt

I reckon there might be many other scenarios where library authors are using `ConditionalWeakTable<TKey, TValue>` as well where the ability to iterate key-value pairs without incurring in extra allocations would be very welcome as well.

## Alternative Designs

While this would still be perfectly safe under normal usage, I understand that it's still technically slightly less safe than the normal enumerator, given that developers could in theory (but also, why?) do weird stuff like copying the enumerator, use it manually outside of a `foreach` block, etc. To that I have two possible answers:

1) The `ConditionalWeakTable<TKey, TValue>` is already in the `System.Runtime.CompilerServices` namespace, which is intrinsically considered "unsafe" and "use at your own risk" in general, so it's assumed that developers using this type (or any of the others in this namespace in general) would know what they're doing. Because of this, I think adding this new API directly on `ConditionalWeakTable<TKey, TValue>` would be fine.
2) In case this proposed API was still considered not safe enough to add to `ConditionalWeakTable<TKey, TValue>` directly, an alternative proposal could be to just move the API to `CollectionMarshal` instead. This would be clunkier (the enumerator type would have to either be moved too or just not be usable directly from `ConditionalWeakTable<TKey, TValue>`, and it'd also need a new `GetEnumerator()` method on it just returning `this`, in order to be usable in a `foreach` block), but in turn it'd entirely separate this slightly more unsafe API into this new special-purposed type (which also contains a bunch of other unsafe APIs for other collection types).

I'd personally argue for 1) given the above (and especially because it'd make the new enumerator much less verbose to use), but at the end of the day I'd be happy with either as long as the core functionality is present, which is true in both cases 😄

## Risks

Very low, as mentioned above. It'd be a niche API on an already niche type, in a niche namespace that's already being used to host all sorts of weird and unsafe APIs. Developers ever finding this should already be very much aware of how to use it properly (especially given that we would also include detailed XML docs to call out all the various potential pitfalls for incorrect usage).

## Additional context

For reference, as mentioned above, we could use this API to completely remove memory allocations in our `WeakReferenceMessenger` implementation in the MVVM Toolkit (part of the Windows Community Toolkit). We particularly focused on speed and efficiency in our messenger types, which are currently far ahead of all other  available MVVM libraries (eg. Prism, Caliburn, etc.). Here's some benchmarks with our current implementations:

```
BenchmarkDotNet=v0.13.0, OS=Windows 10.0.19042.1023 (20H2/October2020Update)
AMD Ryzen 7 2700X, 1 CPU, 16 logical and 8 physical cores
.NET SDK=6.0.100-preview.4.21255.9
  [Host]     : .NET 5.0.6 (5.0.621.22011), X64 RyuJIT
  DefaultJob : .NET 5.0.6 (5.0.621.22011), X64 RyuJIT
```

|            Method |       Mean |     Error | Ratio |      Gen 0 |     Allocated |
|------------------ |-----------:|----------:|------:|-----------:|--------------:|
| **MVVMToolkitStrong** |   **6.428 ms** | 0.0108 ms |  0.08 |          - |             **-** |
|   **MVVMToolkitWeak** |  **20.883 ms** | 0.3085 ms |  0.25 |    31.2500 |     **192,009 B** |
|         MVVMLight |  82.511 ms | 1.5461 ms |  1.00 | 10857.1429 |  45,920,081 B |
|             Prism | 216.294 ms | 2.1281 ms |  2.62 | 18000.0000 |  76,096,445 B |
|     CaliburnMicro | 465.634 ms | 2.7784 ms |  5.65 | 95000.0000 | 397,825,336 B |

This proposed API would allow us to both completely eliminate allocations in the broadcast method for the weak messenger, as well as reducing the performance gap between our two messenger types (cc. @michael-hawker, Windows Community Toolkit lead).

**Comment by michael-hawker** [comment]
Thanks @Sergio0694 for filing this! It's really amazing to see opportunities you've identified in .NET from the work you've done in the Toolkit. Let me know if there's anything I can do to help.

**Comment by Sergio0694** [comment]
Thanks @michael-hawker! Really hoping we can get this in so we can go completely allocation-free on .NET 6 😄
As the next steps here, we'll need an area owner (eg. @joperezr, who I see has already had a look at this proposal) to mark it as ready for review, so that when it passes the live API reviews (assuming it does, of course), I can just open a PR from my branch (I've already implemented these changes in my fork and I'm keeping it on par with `main` for now) 👍
I believe the cutoff date for .NET 6 is around July 17th to merge PRs, so we still have plenty of time to do all this.

## Additional context
Leaving some more info here while I'm at it. The rationale for this proposal is twofold:
- Provide a noticeable memory usage improvement for `ConditionalWeakTable<TKey, TValue>` consumers
- Not introducing any kind of breaking changes or larger concerns with new APIs

With respect to the second point specifically, I'm aware there are two separate proposal to expose `DependentHandle` publicly (ie. https://github.com/dotnet/runtime/issues/19459 and https://github.com/dotnet/runtime/issues/30759), but as expressed by @Maoni0 in a comment ([here](https://github.com/dotnet/runtime/issues/30759#issuecomment-754271405)), there's a concern in exposing `DependentHandle` directly as it'd require GC work to make it efficient enough (and that work is not coming in .NET 6), meaning neither of those proposals really has a chance of making it in .NET 6. With this one instead:
- There is no concern with new public types, as we're just using the existing `ConditionalWeakTable<TKey, TValue>` type
- There are no source/binary breaking changes (the existing enumerators were explicit interface implementations, thankfully)
- The new value-enumerator pattern really just offers a "free" performance/memory improvement to consumers. The fact the proposed type is a `ref struct` means we also don't have to worry about it accidentally being boxed. It would really only be used directly in a `foreach` block, with the compiler producing all the necessary logic to properly dispose it. The only possible way it could cause issues is if developers deliberately used it explicitly and incorrectly, which is (1) against the documentation and (2) we're already in `System.Runtime.*` anyway, which contains a whole bunch of (much more) unsafe APIs and already assumes developers should know how to use APIs in this namespace properly.

This overall just seems like a good improvement for `ConditionalWeakTable<TKey, TValue>` users, with little to no risks and with an easy implementation, without any big API changes nor potentially problematic new types (such as `DependentHandle`) 😊

**Comment by Joe4evr** [comment]
What exactly necessitates that this enumerator be a stack-only struct?

**Comment by Sergio0694** [comment]
@Joe4evr having it be a `ref struct` enables the [disposable `ref struct`-s](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#disposable-ref-structs) feature from C# 8, so that the compiler will generate the `try`/`catch` block with a call to `Dispose` automatically when using a `foreach` statement ([see here](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABADAAgwRhQG4BYAKHIwGY8AmHAYXIG9yd28aMUcBZACgAqAQ2AAbGAB4CWADQ4AlgDsALgD4cK0RICUbDqzIdjOAGbQYwsAAsc/AG7CoOAA7CFz5Zu0w9Rk+yGAewAvvqh5GEUZNR0OCLiUoIA0jAAnvKCAGrCYgCuMGos4Zw4AKJKeQC2MFDCKtA4AOIwKhXVtfXQ/Do4ALwaACYwpsJ5Yiqk/gElsbCmOADOKlB5YCrllTV1DVAlQcGxqWk5+TAACu5QkinpmacFGgx5ULCqOMw4AOatRDhRJWMsWAEAgYj4EHsMAAcjAECoev0NCsClNgqVuDgACIKRYuCCLGCIz5RYxREJAA=)). Without `ref` instead, that `Dispose` method would just be ignored entirely ([see here](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABADAAgwRhQG4BYAKHIwGY8AmHAYXIG9yd28aMUcBZACgAqAQ2AAbGAB4CWADQ4AlgDsALgD4cK0RICUbDqzIdjOAGbQYwsAAsc/AG7CoOAA7CFz5Zu0w9Rk+yGAewAvvqh5GEUZNR0OCLiUoIA0jAAnvKCAGrCYgCuMGos4Zw4AKJKeQC2MFDCKtA4AOIwKhXVtfXQ/Do4ALwaACYwpsJ5Yiqk/gElsQDOKlB5YCrllTV1DVAlQcGxqWk5+TAACu5QkinpmUcFGgx5ULCqOMw4AOatRDhRJcaxwAgEDEfAg9hgADkYAgVD1+hpFgUpsFStwcAARBRzFwQOYwOFvKLGKIhIA==)), and the enumerator would not be disposed, hence breaking the parent `ConditionalWeakTable<TKey, TValue>` instance. Additionally, it also gives us the guarantee that the enumerator can't be boxed, hence reducing the chance of misuse (also boxing this enumerator would defeat the whole point anyway) 🙂

**Comment by stephentoub** [comment]
> having it be a ref struct enables the disposable ref struct-s feature from C# 8, so that the compiler will generate the try/catch block with a call to Dispose automatically when using a foreach statement (see here). Without ref instead, that Dispose method would just be ignored entirely (see here), and the enumerator would not be disposed

With a normal struct you'd just add `: IDisposable`. That C# feature was added for ref structs because they can't implement interfaces.

**Comment by Sergio0694** [comment]
Of course, but given that boxing that would both defeat the point of this enumerator (the existing one works well for that already) while still be more dangerous (eg. it wouldn't have the finalizer, so if you forgot to dispose it yourself you'd still end up breaking the table), I figured making it a `ref struct` would still add some more safety for this specific scenario and just make it better? 🤔

**Comment by stephentoub** [comment]
> but given that boxing that would both defeat the point of this Enumerator

There's no boxing.

**Comment by Sergio0694** [comment]
> "There's no boxing."

That's not what I meant, let me rephrase that 😄
I'm aware that if you have a struct type implementing `IDisposable`, you don't get boxing when used in a `using`/`foreach`, since the `Dispose` call will just be a constrained callvirt and no boxing will be needed, that is fine. What I meant is that given that this enumerator type is inherently less safe than the existing reference type one, making it a `ref struct` just seemed to at least reduce the chance of errors in general (as in, you just couldn't possibly box it by accident if you were to use it directly).

That said, other than that I have no issues with just making it non-`ref` and implementing `IDisposable`, as you said the functionality and codegen would just remain exactly the same when used properly anyway. If someone really wanted to shoot themselves in the foot with this, that'd still be possible in either case anyway, and I don't think the API should worry about those cases too much anyway (especially given in what namespace we're in). I'm happy to change the proposal by adding the interface and removing the `ref` modifier if you prefer 👍

**Comment by Joe4evr** [comment]
> making it a `ref struct` just seemed to at least reduce the chance of errors in general (as in, you just couldn't possibly box it by accident if you were to use it directly).

Most BCL collection types have regular struct enumerators, and I haven't heard anything about them being less safe or error-prone for users, and if they get boxed in favor of convenience (eg. LINQ), so be it. ¯\\\_(ツ)_/¯

**Comment by Sergio0694** [comment]
I never said that `ref struct` enumerators are "less safe" on their own, I was _specifically_ talking about the `ConditionalWeakTable<TKey, TValue>` enumerator, which interacts with the internal reference counting system that the parent type uses to handle containers. This is also the reason why the current reference type enumerator has a finalizer, which is to ensure that the reference count is correctly updated even if the enumerator itself is not properly disposed. In other words, if you eg. grab a `List<T>` enumerator, copy it, box it, throw it away or anything, literally nobody cares. If you did the same with a `ConditionalWeakTable<TKey, TValue>` instance and that enumerator didn't have the finalization logic it currently has, then you'd end up with an inconsistent internal state for the table going forward, like is mentioned in this `Enumerator` comment:

https://github.com/dotnet/runtime/blob/f18f41b14816f02877c721d99ccc761967748184/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/ConditionalWeakTable.cs#L251-L265

That said, I really have no issues whatsoever with updating the proposal, dropping `ref` and adding `IDisposable` 🙂

**Comment by Joe4evr** [comment]
Ah, I see.
![](https://i.imgur.com/TUaf91R.png)

**Comment by stephentoub** [comment]
Taking a step back...

The rest of the CWT implementation is not prepared for this.  Whether or not the Enumerator is a struct or ref struct,  misuse could be very dangerous.  For example,  it would require a very in-depth review to prove to ourselves that the unsafe casts used in the implementation wouldn't end up breaking type/memory safety in the face of misuse that resulted in dependent handles being accessed in unexpected ways. 

I understand the goals here.  I get that your microbenchmarks improve. But having such unsafeness behind one of the most innocuous / ubiquitous API shapes in .NET is problematic.

**Comment by jkotas** [comment]
Are you proposing this enumerator as a workaround for lack of public DependentHandle, or would actually rather see DependentHandle exposed that would allow you to implement your own custom collection in MVVM that does exactly what MVVM needs with no unnecessary overhead?

**Comment by Sergio0694** [comment]
Replying to @stephentoub:

> The rest of the CWT implementation is not prepared for this. Whether or not the Enumerator is a struct or ref struct, misuse could be very dangerous. For example, it would require a very in-depth review to prove to ourselves that the unsafe casts used in the implementation wouldn't end up breaking type/memory safety in the face of misuse that resulted in dependent handles being accessed in unexpected ways.

I wasn't too concerned with those `Unsafe.As<T>(object)` casts as they only seem to be done on the key/value instances, so they don't really seem to be aware of the enumerator types, but yeah I can definitely understand how in general just having those APIs being used in CWT in general means we should be very careful with any changes, for sure 😄

> I get that your microbenchmarks improve.

Your point is still valid, but just to clarify here, those weren't really microbenchmarks as in, just testing a very specific bit of code only traversing the table, but those were full (end-to-end?) benchmarks on our public APIs, which then internally also traversed the CWT instance in use. That's one of the reasons why we'd really like to have _some_ way to improve there, as being able to specifically get rid of all allocations completely in our entire message broadcasting logic would be great both on its own and to also offer a distinguishing feature in comparison with other similar event buffer types from other MVVM libraries.

> But having such unsafeness behind one of the most innocuous / ubiquitous API shapes in .NET is problematic.

That is a fair point, for sure. As an alternative, just as a thought here, would it make sense and/or be better to go with option (2) in the proposal, which is to add an API to grab this enumerator through the `CollectionsMarshal` class instead? Since that type is specifically used to just host a whole bunch of stricty unsafe APIs for various collection types anyway, and this way existing consumers of CWT wouldn't see the new potentially unsafe APIs at all unless they actually went looking for the new API 🤔

Replying to @jkotas:

> Are you proposing this enumerator as a workaround for lack of public DependentHandle, or would you actually rather see DependentHandle exposed that would allow you to implement your own custom collection in MVVM that does exactly what MVVM needs with no unnecessary overhead?

So, the way I was looking at this was that there's two different angles to this.
- For sure, if `DependentHandle` was public, I could just implement my own "domain-specific and lightweight CWT-equivalent type" using that, and just solve all the issues I currently have with CWT, that would definitely be doable. The main issue with this is that it seems to me that type isn't possibly getting exposed any sooner than .NET 7 at the earliest (as per [this comment](https://github.com/dotnet/runtime/issues/30759#issuecomment-754271405)).
- Even without `DependentHandle` being public, my rationale was that I think there's still value in exposing a more efficient and non-allocating way to traverse pairs in a CWT instance, since there'd be plenty of new and existing users that would still want/prefer to use CWT even if `DependentHandle` was public, but that could also benefit from optimizations here.

So basically what I mean is, it'd be great if `DependentHandle` was public and yes it'd also be another way to solve my issue specifically, but I don't see that type being exposed and a proposal such as the one in this issue as being mutually exclusive 🙂

**Comment by jkotas** [comment]
> getting exposed any sooner than .NET 7 at the earliest (as per [this comment](https://github.com/dotnet/runtime/issues/30759#issuecomment-754271405))

This comment basically says that we do not want to encourage more dependent handle uses until the performance is improved. You can use the same argument for delaying any improvements in CWT (such as this proposal) since improvements in CWT are generally going to lead to more dependent handle uses. 

Having said that, I do not agree with this argument. Dependent handle performance is not that bad, dependent handles are never going to free so one will have to be always careful about what they are used for. I think we are doing diservice by not exposing dependent handle as is, and forcing people to use workarounds such as unnecessary complex CWT usage patterns where dependent handle would be much simple solution, or unsupported private reflection hacks.

I think exposing dependent handle would be a lot less contentious than exposing this unsafe enumerator on CWT.

**Comment by Sergio0694** [comment]
I can definitely agree with that, sure. I was indeed confused by the argument for not exposing `DependentHandle` as well, since in my experience that doesn't really reduce the usage as much as just force devs needing that to then do both than and also get the extra overhead and hoops of having to pull in `ConditionalWeakTable<,>` in the process, like I did myself as well 🙂

I'm happy to close this issue then if you think that just making `DependentHandle` public would be just better overall.

I guess the question now becomes: do you reckon it'd be possible to get `DependentHandle` public in .NET 6? And if so, is there anything we could do to help with that? Even just opening a new proposal specifically for that maybe, and linking this one? 😄

**Comment by jkotas** [comment]
https://github.com/dotnet/runtime/issues/19459 is existing proposal to expose DependentHandle with some open questions. We can iterate on that, no need to create new issue.

**Comment by Sergio0694** [comment]
Sure thing! Was just double checking since that proposal was slightly different than just exposing `DependentHandle`.
Closing this issue and crossing my fingers for https://github.com/dotnet/runtime/issues/19459 then! 😄
